bwe/BandwidthEstimator.cpp

```diff
@@ -31,164 +43,243 @@ BandwidthEstimator::BandwidthEstimator(const Config& config)
-    _covarianceP = statePredictionCovariance - kalmanGain * covDelay * math::transpose(kalmanGain);
+    _covarianceP = statePredictionCovariance - crossCovariance * math::transpose(kalmanGain);
+
+    sanitizeState(observedDelay, packetSize * 8, _state);
+
+    assert(math::isValid(_covarianceP));
     math::makeSymmetric(_covarianceP);
     assert(math::isSymmetric(_covarianceP));
     assert(math::isValid(_covarianceP));
 
     _observedDelay = observedDelay;
-    _state(0) = std::max(0.0, std::min(_state(0), _config.maxNetworkQueue * 8));
 
-    if (_state(2) > observedDelay + 12.0)
+    _previousReceiveTime = receiveTimeNs;
+    _previousTransmitTime = transmitTimeNs;
+}
+
+void BandwidthEstimator::sanitizeState(const double observedDelay,
+    const double packetBits,
+    math::Matrix<double, 3>& state)
+{
+    state(Bandwidth) = math::clamp(state(Bandwidth), _config.modelMinBandwidth, _config.estimate.maxKbps);
+
+    state(QueuedBits) = std::max(packetBits, state(QueuedBits));
+
+    if (observedDelay - predictAbsoluteDelay(state) < 0 && state(QueuedBits) > _config.mtu * 3)
     {
-        // under estimating delay. Likely in congestion and over estimating bw
-        _congestion.margin = _config.congestion.backOff;
-        _state(1) = std::min(_state(1), _receiveBandwidth.get(550 * utils::Time::ms) * utils::Time::ms);
+        // adjust queue if we received packet earlier than expected. If bw is higher than expected, queue should also
+        // have drained more
+        const double delayErr = predictAbsoluteDelay(state) - observedDelay;
+        state(QueuedBits) -= delayErr * state(Bandwidth) / 3;
+        state(QueuedBits) = std::max(packetBits, state(QueuedBits));
     }
-    _state(2) = std::min(observedDelay, _state(2));
-    _state(1) = std::min(_config.estimate.maxKbps, std::max(_config.estimate.minKbps, _state(1)));
 
-    updateCongestionMargin(
-        utils::Time::diff(_previousReceiveTime, receiveTimeNs) / static_cast<double>(utils::Time::ms));
+    state(QueuedBits) = math::clamp(state(QueuedBits), packetBits, _config.maxNetworkQueue * 8);
 
-    _previousReceiveTime = receiveTimeNs;
-    _previousTransmitTime = transmitTimeNs;
+    state(ClockOffset) = std::min(observedDelay, state(ClockOffset));
+}
+
+void BandwidthEstimator::calculateProcessNoise(const math::Matrix<double, 3>& currentState,
+    const double actualDelay,
+    const double observationError,
+    const uint32_t packetSize,
+    const uint64_t receiveTimeNs,
+    math::Matrix<double, 3>& processNoise,
+    double& measurementNoise)
+{
+    measurementNoise = _config.measurementNoise;
+
+    const double longerQueue = _config.mtu * 8 * 2;
+    if (_congestion.consecutiveOver == 0)
+    {
+        _congestion.estimateBeforeCongestion = currentState(Bandwidth);
+    }
+
+    if (_congestion.consecutiveUnder > 15 || _congestion.consecutiveOver > 30)
+    {
+        // adapt faster at consistent observed low delay or prolonged observed higher delay
```

Comment:
If we observe lower delay consistently we should look for lower bandwidth and adjust faster to observation.
If we observe higher delay than expected we wait a bit longer before adjusting bw.
We adjust more, the longer this has been observed.